Sources/PackageLoading/ToolsVersionLoader.swift

```diff
@@ -88,32 +88,122 @@ extension Manifest {
+                let unicodeCodePointsOfCharactersPrefixedByUPlus: [String] = unicodeCodePointsOfCharacters.map { codePoint in
+                    var codePointString = String(codePoint, radix: 16).uppercased()
+                    if codePointString.count < 4 {
+                        codePointString = String(repeating: "0", count: 4 - codePointString.count) + codePointString
+                    }
+                    return "U+\(codePointString)"
+                }
+                // FIXME: Use `ListFormatter` instead?
+                return "[\(unicodeCodePointsOfCharactersPrefixedByUPlus.joined(separator: ", "))]"
+            }
+            
             switch self {
-            case .inaccessiblePackage(let packageDir, let reason):
-                return "the package at '\(packageDir)' cannot be accessed (\(reason))"
-            case .inaccessibleManifest(let manifestFile, let reason):
-                return "the package manifest at '\(manifestFile)' cannot be accessed (\(reason))"
-            case .malformedToolsVersion(let versionSpecifier, let currentToolsVersion):
+            case let .inaccessiblePackage(packageDirectoryPath, reason):
+                return "the package at '\(packageDirectoryPath)' cannot be accessed (\(reason))"
+            case let .inaccessibleManifest(manifestFilePath, reason):
+                return "the package manifest at '\(manifestFilePath)' cannot be accessed (\(reason))"
+            case let .nonUTF8EncodedManifest(manifestFilePath):
+                return "the package manifest at '\(manifestFilePath)' cannot be decoded using UTF-8"
+            case let .malformedToolsVersion(versionSpecifier, currentToolsVersion):
                 return "the tools version '\(versionSpecifier)' is not valid; consider using '// swift-tools-version:\(currentToolsVersion.major).\(currentToolsVersion.minor)' to specify the current tools version"
+            case let .malformedToolsVersionSpecification(malformation):
+                switch malformation {
+                case let .label(label):
+                    return "the tools version specification label '\(label)' is malformed; consider using '// swift-tools-version:\(ToolsVersion.currentToolsVersion)' to specify the current tools version"
+                case let .versionSpecifier(versionSpecifier):
+                    return "the tools version '\(versionSpecifier)' is not valid; consider using '// swift-tools-version:\(ToolsVersion.currentToolsVersion)' to specify the current tools version"
+                case let .entireLine(line):
+                    return "the tools version specification '\(line)' is not valid; consider using '// swift-tools-version:\(ToolsVersion.currentToolsVersion)' to specify the current tools version"
+                }
+            // FIXME: The error messages probably can be more concise, while still hitting all the key points.
```

Comment:
I think that can be revised over time.  Some of SwiftPM's current diagnostics are far too brief, and the detail can be useful.

Reply:
I removed this FIXME in d82202b88999fc6ce7f076e7e27d7bf72374c471.